libutils/threaded_queue.c

```diff
@@ -0,0 +1,393 @@
+{
+    assert(queue != NULL);
+
+    ThreadLock(queue->lock);
+    size_t const count = queue->size;
+    ThreadUnlock(queue->lock);
+
+    return count;
+}
+
+size_t ThreadedQueueCapacity(ThreadedQueue const *queue)
+{
+    assert(queue != NULL);
+
+    ThreadLock(queue->lock);
+    size_t const capacity = queue->capacity;
+    ThreadUnlock(queue->lock);
+
+    return capacity;
+}
+
+bool ThreadedQueueIsEmpty(ThreadedQueue const *queue)
+{
+    assert(queue != NULL);
+
+    ThreadLock(queue->lock);
+    bool const empty = (queue->size == 0);
+    ThreadUnlock(queue->lock);
+
+    return empty;
+}
+
+ThreadedQueue *ThreadedQueueCopy(ThreadedQueue *queue)
+{
+    assert(queue != NULL);
+
+    ThreadLock(queue->lock);
+
+    ThreadedQueue *new_queue = xmemdup(queue, sizeof(ThreadedQueue));
+    new_queue->data = xmalloc(sizeof(void *) * queue->capacity);
+    memcpy(new_queue->data, queue->data,
+           sizeof(void *) * new_queue->capacity);
+
+    ThreadUnlock(queue->lock);
+
+    // enables error checking for deadlocks
+    pthread_mutexattr_t attr;
+    pthread_mutexattr_init(&attr);
+    int ret = pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_ERRORCHECK);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to use error-checking mutexes for queue, "
+            "falling back to normal ones (pthread_mutexattr_settype: %s)",
+            GetErrorStrFromCode(ret));
+        pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_NORMAL);
+    }
+
+    new_queue->lock = xmalloc(sizeof(pthread_mutex_t));
+    ret = pthread_mutex_init(new_queue->lock, &attr);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to initialize mutex (pthread_mutex_init: %s)",
+            GetErrorStrFromCode(ret));
+        pthread_mutexattr_destroy(&attr);
+        free(new_queue->lock);
+        free(new_queue);
+        return NULL;
+    }
+
+    new_queue->pcond = xmalloc(sizeof(pthread_cond_t));
+    ret = pthread_cond_init(new_queue->pcond, NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to initialize thread condition "
+            "(pthread_cond_init: %s)",
+            GetErrorStrFromCode(ret));
+        free(new_queue->lock);
+        free(new_queue->pcond);
+        free(new_queue);
+        return NULL;
+    }
+
+    return new_queue;
+}
+
+/**
+  @brief Destroys data in range.
+  @warning Assumes that locks are acquired.
+  @param [in] queue Pointer to struct.
+  @param [in] start Start position to destroy from.
+  @param [in] end Where to stop.
+  */
+static void _DestroyRange(ThreadedQueue *queue, size_t start, size_t end)
+{
+    assert(queue != NULL);
+    if (start > queue->capacity || end > queue->capacity)
+    {
+        return;
+    }
+
+    if (queue->ItemDestroy && queue->size > 0)
+    {
+        queue->ItemDestroy(queue->data[start]);
+
+        // in case start == end, start at second element in range
+        for (size_t i = start + 1; i != end; i++)
```

Comment:
Start is destroyed outside the loop. I do it this way since `queue->tail` can be the same as `queue->head`. I think it makes more sense that end is excluded, since it then makes sense to pass in just `queue->head` and `queue->tail`.